### PR TITLE
implement new chaosctl

### PR DESCRIPTION
**Pull request by Hexilee**
Signed-off-by: hexilee <i@hexilee.me>

### What problem does this PR solve?

The PR #2192 brings us a ctrl server, now we are going to implement a new `chaosctl` based on it.

Problem Summary:

### What is changed and how it works?

Proposal: [rfc#19](https://github.com/chaos-mesh/rfcs/pull/19) 

What's Changed:

### Related changes

* PR to update `chaos-mesh/website`/`chaos-mesh/website-zh`:
* Need to update Chaos Dashboard component, related issue:
* Need to cheery-pick to the release branch

### Checklist

Tests

- [ ] Unit test
- [x] E2E test
- [x] Manual test (add detailed scripts or steps below)

**New Features**

#### Basic Usage

```bash
bin/chaosctl query <resources>
```

For example, use the following command to query the name of chaos daemons and the name of PodHttpChaos instances owned by HttpChaos `http-chaos-example`.

```bash
bin/chaosctl query component:daemon/name http:http-chaos-example/podhttp/name
```

#### Namespace Scope

The `-n` flag specifies the namespace scope of the root resources. For example, use the following command to query the name of chaos daemons in the `chaos-testing` namespace.

```bash
bin/chaosctl query component:daemon/name -n chaos-testing
```

#### Root Resource

Sometimes, all resources we need are sub-resources of a root resource. Following two commands get the same result.

```bash
bin/chaosctl query component:daemon/name component:daemon/annotations
```

```bash
bin/chaosctl query -r component:daemon name annotations
```

> For a better completion experience, the resources should be placed behind the `-r` flag.

#### Field List

Sometimes, some resources we need are sub-resources of a root resource, but others are not, we can use the field list to list those resources that have the same parent.

Following two commands get the same result.

```bash
bin/chaosctl query component:daemon/name component:daemon/annotations http:http-chaos-example/podhttp/name
```

```bash
bin/chaosctl query component:daemon/name,annotations http:http-chaos-example/podhttp/name
```

#### Resource List

If you want to list all resources like `HttpChaos`, just use tag `@all` behind resource name.

```bash
bin/chaosctl query http@all/podhttp/name
```

- [ ] No code

Side effects

- [ ] Breaking backward compatibility

### Release note 

```release-note
Please add a release note.
If you don't think this PR needs a release note then fill it with None.
```

### CODE REVIEW

**Review by STRRL — COMMENTED** [review]
reviewing in progress, I will appending more comments soon :D

**Comment by STRRL** [comment]
the path of chaos resource is mostly like `io/<chaos-name>/<field>` for fetching certain field of certain IOChaos, and `ios` for list all the IOChaos. I think using abbr. (like) `ios` would bring confusion. (and there are no resource for PodChaos, because `pod` stands for the real kubernetes Pod🤣)

Use `IOChaos`(case-seneitive) is much better. I known that will change the schema and introduce more lines in thies PR, but I think that's kind of necessary. Please think about it. :heart:

**Comment by Hexilee** [comment]
> the path of chaos resource is mostly like `io/<chaos-name>/<field>` for fetching certain field of certain IOChaos, and `ios` for list all the IOChaos. I think using abbr. (like) `ios` would bring confusion. (and there are no resource for PodChaos, because `pod` stands for the real kubernetes Pod🤣)
> 
> Use `IOChaos`(case-seneitive) is much better. I known that will change the schema and introduce more lines in thies PR, but I think that's kind of necessary. Please think about it. ❤️

Sounds reasonable, but the auto-completion relies on the default mapping (io <--> ios) to list all resources. Maybe we should provide a mapping API on the ctrlserver, then we can rename the path.

File: pkg/chaosctl/cmd/get.go [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+		if root != "" {
+			prefix = append(prefix, ctrlclient.StandardizeQuery(root)...)
+		}
+		return prefix
+	}
+
+	// getCmd represents the get command
+	var getCmd = &cobra.Command{
+		Use:   "get [resources]",
+		Short: "get the target resources",
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/cmd/get.go]: Fixed

File: pkg/chaosctl/cmd/schema.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
+
+	ctrlclient "github.com/chaos-mesh/chaos-mesh/pkg/ctrl/client"
+)
+
+// schemaCmd represents the schema command
+var schemaCmd = &cobra.Command{
+	Use:   "schema",
+	Short: "get the graphql schema of ctrl server",
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/cmd/schema.go]: Fixed

**Comment by Hexilee** [comment]
> image The format of the first word should be standardised

Fixed.

File: pkg/ctrl/client/forward.go [hunk 3]
```diff
@@ -11,12 +11,22 @@
...
+	"github.com/pkg/errors"
 	"k8s.io/cli-runtime/pkg/genericclioptions"
 	"k8s.io/client-go/rest"
 	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	"github.com/chaos-mesh/chaos-mesh/pkg/portforward"
+)
+
+const (
+	ManagerSvc     = "svc/chaos-mesh-controller-manager"
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/ctrl/client/forward.go]: Done

File: go.mod [hunk 4]
```diff
@@ -37,10 +37,12 @@ require (
 	github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e // indirect
 	github.com/golang/mock v1.5.0
 	github.com/golang/protobuf v1.4.3
+	github.com/google/uuid v1.3.0 // indirect
 	github.com/gorilla/mux v1.7.0 // indirect
 	github.com/grpc-ecosystem/go-grpc-middleware v1.2.0
 	github.com/grpc-ecosystem/go-grpc-prometheus v1.2.0
 	github.com/hashicorp/golang-lru v0.5.5-0.20210104140557-80c98217689d
+	github.com/hasura/go-graphql-client v0.2.0
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply go.mod:36]: ignore

**Comment by Hexilee** [comment]
New progress:
- the debug and logs subcommands are completely reimplemented based on the ctrl server
- all messages exposed to users are printed by `PrettyPrint` instead of the logger
- add more integration tests

PTAL again! @STRRL @YangKeao @cwen0

**Comment by STRRL** [comment]
Hi @Hexilee , please resolve the conflicts

**Comment by Hexilee** [comment]
> Hi @Hexilee , please resolve the conflicts

Resolved

File: pkg/chaosctl/debug/networkchaos/debug.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+package networkchaos
+
+import (
+	"context"
+
+	"github.com/hasura/go-graphql-client"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	"github.com/chaos-mesh/chaos-mesh/pkg/chaosctl/common"
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/debug/networkchaos/debug.go:24]: Fixed

**Comment by Hexilee** [comment]
I will reopen this PR after #2950 is merged.

**Pull request closed without merging**